internal/sanitize/sanitize.go

```diff
@@ -0,0 +1,329 @@
+		}
+		newImageDestination = f.Name()
+		defer func() {
+			if tw != nil {
+				if err := tw.Close(); err != nil {
+					logrus.Warnf("wrapping up writing copy of base image to archive %q: %v", newImageDestination, err)
+				}
+			}
+			if f != nil {
+				if err := f.Close(); err != nil {
+					logrus.Warnf("closing copy of base image in archive file %q: %v", newImageDestination, err)
+				}
+			}
+		}()
+		// archive only the archive file for copying to the new archive file
+		imageArchive, err = newSingleItemArchive(contextDir, archiveSource)
+		isEmbeddedArchive = true
+		// generate the new reference using the temporary file's name
+		newFrom = transportName + ":" + strings.Join(append([]string{newImageDestination}, refLeftover...), ":")
+	case ociLayoutTransport.Transport.Name(): // this is a directory tree
+		// this takes the form path[:stuff]
+		transportRef := restOfImageName
+		parts := strings.Split(transportRef, ":")
+		archiveSource = parts[0]
+		refLeftover := parts[1:]
+		// create a new directory to use as our new layout directory
+		if newImageDestination, err = newDirectoryDestination(tmpdir); err != nil {
+			return "", fmt.Errorf("creating temporary copy of base image: %w", err)
+		}
+		// archive the entire layout directory for copying to the new layout directory
+		tarOptions := &archive.TarOptions{}
+		imageArchive, err = chrootarchive.Tar(filepath.Join(contextDir, archiveSource), tarOptions, contextDir)
+		// generate the new reference using the directory
+		newFrom = transportName + ":" + strings.Join(append([]string{newImageDestination}, refLeftover...), ":")
+	case directoryTransport.Transport.Name(): // this is also a directory tree
+		// this takes the form of just a path
+		transportRef := restOfImageName
+		// create a new directory to use as our new image directory
+		if newImageDestination, err = newDirectoryDestination(tmpdir); err != nil {
+			return "", fmt.Errorf("creating temporary copy of base image: %w", err)
+		}
+		// archive the entire directory for copying to the new directory
+		archiveSource = transportRef
+		tarOptions := &archive.TarOptions{}
+		imageArchive, err = chrootarchive.Tar(filepath.Join(contextDir, archiveSource), tarOptions, contextDir)
+		// generate the new reference using the directory
+		newFrom = transportName + ":" + newImageDestination
+	default:
+		return "", fmt.Errorf("unexpected container image transport %q", transportName)
+	}
+	if err != nil {
+		return "", fmt.Errorf("error archiving source at %q under %q", archiveSource, contextDir)
+	}
+
+	// start reading the archived content
+	defer func() {
+		if err := imageArchive.Close(); err != nil {
+			logrus.Warn(err)
+		}
+	}()
+	tr := tar.NewReader(imageArchive)
+	hdr, err := tr.Next()
+	for hdr != nil {
+		// if the archive we're parsing is expected to have an archive as its only item,
+		// assume it's the first (and hopefully, only) item, and switch to stepping through
+		// it as the archive
+		if isEmbeddedArchive {
```

Comment:
(Sanity-check that the item is a regular file? Probably unnecessary, the tests would notice if the output of `newSingleItemArchive` changed.)